### PR TITLE
feat(app-router): support useRouter bfcacheId semantics

### PR DESCRIPTION
**Pull request by manNomi**
## Summary
- Implements segment-scoped `useRouter().bfcacheId` semantics for the App Router shim.
- Persists Vinext-owned bfcache id maps in history state so back/forward traversal can restore previous ids.
- Provides bfcache contexts through the browser/SSR route tree and layers the contextual `bfcacheId` onto the mounted `AppRouterContext` router.
- Preserves bfcache metadata during hash-only navigation while avoiding stale scroll restoration/custom history state.
- Includes two browser redirect lifecycle fixes that affect App Router navigations even when user code does not read `bfcacheId`; these are called out below.

## Feedback addressed
This is a fresh follow-up based on the review feedback from the earlier bfcacheId PR. In particular, this version:
- keeps `AppRouterContext` as the authority for router methods and only adds the contextual `bfcacheId` in `useRouter()`
- formats the public zero sentinel as `_b_0_` while sharing the internal zero sentinel constant across the shim/runtime
- documents the intentional raw `"0"` hydration sentinel inside `BfcacheIdMap`; minted ids use `_b_N_`, and the public hook formats both into the opaque public shape
- documents the intentional hard-reload reset to the zero sentinel and uses version-gated history restoration to avoid cross-document id collisions
- preserves bfcache ids after reducer-level element merges, including planner-approved parallel route state
- clears stale restored ids across redirected traversals and server-action/cache invalidation paths
- includes active slot route identity for intercepted/parallel slot bfcache ids
- adds programmatic `router.push()` / `router.replace()` and forward-after-reload coverage in the E2E fixture
- documents that Next.js Activity-backed form-state-on-back assertions are intentionally omitted until vinext implements React Activity
- verifies same-segment input state survives search-param navigation while the leaf `bfcacheId` stays stable

## Compatibility note
`useRouter()` now returns a memoized wrapper around the mounted context router instead of the old module-level singleton. This is intentional Next.js App Router parity so custom/instrumented context router methods are preserved while the nearest segment's `bfcacheId` is added. This is slightly more permissive than Next.js's explicit method list, but it avoids dropping vinext-provided or user-instrumented router methods.

## Redirect lifecycle fixes included in this PR
These fixes are bundled here because redirect hops need to keep bfcache metadata coherent, but they are genuine browser navigation behavior changes for all App Router users:

- Cached visited-response redirects are now followed. Previously, a cached route whose stored response pointed at a redirect target could be treated as the terminal payload. The browser navigation loop now resolves that redirect lifecycle hop, applies the max-redirect guard, and continues to the final target.
- Flight redirects now realign history bookkeeping. Previously, the `X-Vinext-Rsc-Redirect` path updated `currentHref` and the redirect count, but could leave `currentHistoryMode` and `currentPrevNextUrl` stale. The redirect path now defaults unspecified redirects to `replace` and threads the request's previous/next URL through the redirect chain, preventing redirected navigations from committing with the wrong history mode or previous/next metadata.
- For traversal navigations, every redirect hop that changes the target href clears `restoredBfcacheIds`. This keeps vinext's `restored > currentValue > mint` priority chain from applying stale history-state ids from the pre-redirect entry to the redirected payload.

## Tests
- `pnpm exec vp check packages/vinext/src/server/app-bfcache-id.ts packages/vinext/src/server/app-browser-entry.ts packages/vinext/src/server/app-browser-state.ts packages/vinext/src/shims/navigation.ts packages/vinext/src/shims/slot.tsx tests/shims.test.ts tests/app-browser-entry.test.ts tests/e2e/app-router/nextjs-compat/use-router-bfcache-id.spec.ts`
- `pnpm exec vp test run --project unit tests/shims.test.ts -t "useRouter|hash-only"`
- `pnpm exec vp test run --project unit tests/app-browser-entry.test.ts -t "bfcache|history state|version|intercepted slot|redirected traverse"`
- `pnpm exec vp check tests/e2e/app-router/nextjs-compat/use-router-bfcache-id.spec.ts`
- `PLAYWRIGHT_PROJECT=app-router pnpm run test:e2e tests/e2e/app-router/nextjs-compat/use-router-bfcache-id.spec.ts`

### CODE REVIEW

**Comment by NathanDrake2406** [comment]
@codex review

File: packages/vinext/src/server/app-browser-state.ts [hunk 1]
```diff
@@ -135,6 +141,137 @@ type NonDispatchPendingNavigationCommitDispositionDecision = {
...
+  const parsed = AppElementsWire.parseElementKey(id);
+  if (!parsed) return null;
+
+  if (parsed.kind === "page") {
+    return `${id}@${pathname}`;
+  }
+
+  if (parsed.kind === "layout" || parsed.kind === "slot" || parsed.kind === "template") {
+    const segmentCount = getVisibleTreePathSegments(parsed.treePath).length;
+    return `${id}@${getPathPrefix(pathname, segmentCount)}`;
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/server/app-browser-state.ts]: I’ve updated it accordingly.

**Comment by NathanDrake2406** [comment]
Important PR bro, I think it’ll get merged when maintainer is online. That said, there are a few improvements to be made.
Request changes: architecture is good, but I think we should tighten two seams before merging.
## Blocking
### 1. `useRouter()` should preserve the actual `AppRouterContext` router
Right now `useRouter()` reads `AppRouterContext`, checks that it exists, then ignores the context value and returns a clone of the module-level `_appRouter`.
That breaks context ownership. The provider should remain the authority for router methods; `useRouter()` should only add the contextual `bfcacheId`.
This also better matches Next’s shape: read the router from context, read the nearest segment/cache identity, then return the router methods plus the contextual `bfcacheId`.
Suggested shape:
```ts
export function useRouter() {
  if (!AppRouterContext || typeof React.useContext !== "function") {
    throw new Error("invariant expected app router to be mounted");
  }
  const router = React.useContext(AppRouterContext);
  if (router === null) {
    throw new Error("invariant expected app router to be mounted");
  }
  const bfcacheId = readBfcacheIdFromContext();
  return React.useMemo(
    () => ({
      ...router,
      bfcacheId,
    }),
    [router, bfcacheId],
  );
}

```

That removes the need for the global _appRouterByBfcacheId LRU and avoids silently dropping custom/instrumented router methods later.

2. Public bfcacheId format should match the observable Next shape

The PR currently treats initial hook-visible IDs as "0". Next’s public hook materialises the zero cache node as "_b_0_".

Even though the value is opaque, this PR is Next-compat, and the _b_${n}_ format exists to make the value safe to concatenate into user keys. I’d prefer either storing _b_0_ in the map or storing internal numeric IDs and formatting at useRouter().

Architecture hardening worth doing now

I would not block on a full Route Manifest/semantic graph rewrite in this PR. That would make this slice too big.

But I do think we should create the seam now:

* Put all bfcache segment identity derivation behind one helper/module, e.g. createBfcacheSegmentIdentity(...).
* Document that the current AppElementsWire.parseElementKey(...) + pathname prefix logic is a legacy bridge, not the final semantic authority.
* Do not let wire-key parsing spread beyond that helper.
* Make history state strictly typed read/write transport for restoring prior IDs; it should not become route meaning.
* Add a test that a custom AppRouterContext provider keeps its custom methods after useRouter() adds bfcacheId.

This keeps the PR small while preventing the architecture from drifting back into “flat wire keys are the router’s brain”.

Test gaps

The new tests are useful, especially around push/replace and reducer preservation. I’d still add:

1. Browser back preserves actual keyed client state, not just the restored ID.
2. Server-action refresh preserves state, since this PR touches the same-URL/server-action commit path.
3. useRouter() with a non-singleton context router preserves the context router’s methods.

After those changes, I think the architecture is solid and aligned with the navigation direction.

**Comment by NathanDrake2406** [comment]
Blocking issue:

`BrowserRoot` currently initialises `bfcacheIds` with `createInitialBfcacheIdMap(resolvedElements)`, then the initial layout effect writes that map back into the current `history.state`.

That loses a persisted bfcache map when hydrating after a hard reload of a non-initial session-history entry.

Failure shape:

1. `/x/1` starts with `_b_0_`.
2. Push `/x/2`; leaf gets a fresh `_b_n_`.
3. Hard reload on `/x/2`.
4. Hydration resets the current entry’s map to `"0"` and persists it.
5. Browser back can now collide with `/x/1`’s `_b_0_`, breaking the “fresh push gets fresh identity, traversal restores old identity” invariant.

I think initial browser state should seed from `readHistoryStateBfcacheIds(window.history.state)` when present, filtered to the current rendered element/layout ids, and only fall back to `createInitialBfcacheIdMap(resolvedElements)` when no valid persisted map exists. Restored `_b_n_` values should also advance the mint counter so future ids cannot collide.

Test I’d add before merge:

- Navigate `/x/1 -> /x/2`.
- Capture both bfcache ids.
- Hard reload on `/x/2`.
- Assert `/x/2` keeps its pre-reload id.
- Type into the `/x/2` leaf input.
- Browser back to `/x/1`.
- Assert `/x/1` restores its old id and does not inherit `/x/2`’s leaf input state.

I’d also port Next’s original browser-back form-state assertion, not just the restored-id assertion. The id text proves the map changed; the form-state assertion proves the user-visible semantics are correct.
Link: https://github.com/vercel/next.js/blob/56d95137fd6d84f4bc1e5ef2bb31e0136d5fad9c/test/e2e/app-dir/use-router-bfcache-id/use-router-bfcache-id.test.ts

**Comment by NathanDrake2406** [comment]
@codex review

**Comment by NathanDrake2406** [comment]
I think slots/intercepted routes still have an identity bug.

`createBfcacheSegmentIdentity()` treats slots like layouts/templates: id + pathname prefix. But slot ids are owner-stable, e.g. `slot:modal:/feed`, while the active slot page can change underneath it.
So `/photo/1` intercepted into `@modal` → `/photo/2` intercepted into the same `@modal` could preserve the same `bfcacheId`, even though it’s a fresh slot page. 

Can we make slot identity include the active slot route/page identity, and add an E2E for fresh intercepted-slot navigation resetting state, while browser back restores it?

**Comment by manNomi** [comment]
@james-elicx Would it be okay to trigger another /bigbonk review ?

**Comment by manNomi** [comment]
@james-elicx Would it be okay to trigger one final /bigbonk review?

File: packages/vinext/src/shims/slot.tsx [hunk 2]
```diff
@@ -119,6 +121,12 @@ function warnTransportMetadataEntry(id: string): void {
   console.warn("[vinext] Transport metadata value found under App Router render entry: " + id);
 }
 
+function BfcacheSlotBoundary({ content, id }: { content: React.ReactNode; id: string }) {
+  const SegmentContext = BfcacheSegmentIdContext;
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/shims/slot.tsx:125]: Already addressed: `BfcacheSlotBoundary` now reads the context into a local `SegmentContext` and returns the content unchanged when the context is unavailable, avoiding the non-null assertion.

File: packages/vinext/src/server/app-browser-state.ts [hunk 3]
```diff
@@ -135,6 +142,196 @@ type NonDispatchPendingNavigationCommitDispositionDecision = {
...
+  try {
+    return AppElementsWire.readMetadata(elements);
+  } catch {
+    return null;
+  }
+}
+
+function createActiveSlotIdentity(id: string, metadata: AppElementsMetadata | null): string | null {
+  const activeSlotBinding = metadata?.slotBindings.find((binding) => binding.slotId === id);
+  if (activeSlotBinding?.activeRouteId != null && activeSlotBinding.activeRouteId !== "") {
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/server/app-browser-state.ts]: Already addressed: `createActiveSlotIdentity` now uses a strict guard for `activeRouteId` (`!= null` plus an empty-string check) before falling back to interception metadata.

File: packages/vinext/src/server/app-browser-entry.ts [hunk 4]
```diff
@@ -808,21 +881,28 @@ function BrowserRoot({
...
-    visibleCommitVersion: 0,
-  });
+  const [treeStateValue, setTreeStateValue] = useState<AppRouterState | Promise<AppRouterState>>(
+    () => ({
+      activeOperation: null,
+      // A hard reload starts a new browser document without Next.js's
+      // in-memory CacheNode/Activity tree. Hydrate the new document on the
+      // zero sentinel and rely on the document-scoped bfcache version gate to
+      // reject stale ids persisted by previous documents.
+      bfcacheIds: createInitialBfcacheIdMap(resolvedElements),
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/server/app-browser-entry.ts]: Addressed in `55135824`: the hydration comment now explicitly calls this intentional Next.js parity and explains that hard reload starts a new browser document on the zero sentinel while the document-scoped bfcache version gate rejects stale ids from previous documents.

File: packages/vinext/src/shims/navigation.ts [hunk 5]
```diff
@@ -1572,6 +1606,30 @@ const _appRouter = {
   },
 };
 
+function formatPublicBfcacheId(value: string | null | undefined): string {
+  if (!value || value === "0") return "_b_0_";
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/shims/navigation.ts]: Addressed in `55135824` by moving the internal zero sentinel into shared bfcache id constants and using `INITIAL_BFCACHE_ID` / `PUBLIC_INITIAL_BFCACHE_ID` in the navigation shim, so the public formatting can no longer drift from the runtime sentinel.

File: packages/vinext/src/server/app-history-state.ts [hunk 6]
```diff
@@ -55,6 +66,25 @@ export function createHistoryStateWithNavigationMetadata(
...
+  if (metadata.bfcacheIds !== undefined) {
+    if (bfcacheIdsWereCleared) {
+      delete nextState[VINEXT_BFCACHE_IDS_HISTORY_STATE_KEY];
+      delete nextState[VINEXT_BFCACHE_VERSION_HISTORY_STATE_KEY];
+    } else {
+      nextState[VINEXT_BFCACHE_IDS_HISTORY_STATE_KEY] = { ...metadata.bfcacheIds };
+    }
+  }
+
+  if (metadata.bfcacheVersion !== undefined) {
```
[the remark this answers is not in the document]
Comment (manNomi) [reply packages/vinext/src/server/app-history-state.ts:84]: Already addressed: `createHistoryStateWithNavigationMetadata` now tracks `bfcacheIdsWereCleared` and skips/re-clears the version field when ids are explicitly cleared. The regression is covered by the empty `bfcacheIds` unit test.

**Comment by manNomi** [comment]
@james-elicx could you run /bigbonk review once more when you have a chance?

**Comment by manNomi** [comment]
Addressed the minor duplication by extracting the repeated history-state sync predicate into `isHistoryStateNavigationMetadataInSync(...)`.

No behavior change intended: the helper keeps the same `previousNextUrl`, optional `bfcacheIds`, map equality, and current-version checks, and both the early return and post-`replaceState` verification now call the same predicate.

**Comment by manNomi** [comment]
Addressed the functional concern around the `segmentId` truthiness check.

`readBfcacheIdFromContext` now uses a strict nullish check (`segmentId !== null`) before looking up the bfcache id map, so an empty string segment id is treated as a valid context key instead of falling back to the initial sentinel. I also added a focused `useRouter()` unit test covering the empty-string context key case.

Targeted local verification passed:

```sh
pnpm exec vp check packages/vinext/src/shims/navigation.ts tests/shims.test.ts
pnpm exec vp test run --project unit tests/shims.test.ts -t "useRouter"
```

**Comment by james-elicx** [comment]
/bigbonk re-review for issues and next.js parity

**Comment by james-elicx** [comment]
/bigbonk re-review for issues and next.js parity

**Comment by james-elicx** [comment]
/bigbonk re-review for issues and next.js parity

**Comment by manNomi** [comment]
Addressed the latest Bonk feedback and rebased onto latest main. Targeted checks are green.
